From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: jho5245 <jho5245@naver.com>
Date: Thu, 18 Sep 2025 12:08:14 +0900
Subject: [PATCH] Edit Ingredient 2


diff --git a/net/minecraft/world/item/crafting/Ingredient.java b/net/minecraft/world/item/crafting/Ingredient.java
index cd1762e9e8c124a1e08c9551f4782b4ecf55b667..6b2bda82b5e33ffaf19326d1a52f8d9004333bdb 100644
--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -36,7 +36,11 @@ public final class Ingredient implements StackedContents.IngredientInfo<io.paper
     // CraftBukkit start
     @javax.annotation.Nullable
     private java.util.Set<ItemStack> itemStacks; // Paper - Improve exact choice recipe ingredients
+<<<<<<< Updated upstream
     public Predicate<ItemStack> stackPredicate; // Cucumper - Add PredicateChoice
+=======
+    public Predicate<ItemStack> stackPredicate;
+>>>>>>> Stashed changes
 
     public boolean isExact() {
         return this.itemStacks != null;
